code/obj/item/implant.dm

```diff
@@ -890,6 +896,366 @@ ABSTRACT_TYPE(/obj/item/implant/revenge)
+	/// The value slowly ramps up over time, but is reset upon being activated. This makes short-term overuse very punishing,
+	/// but allows it to recover decently quickly if given time to rest.
+	var/heat_dissipation = 1
+	/// If `heat` is above this value, each activation has a chance to break the implant permanently.
+	var/const/heat_danger_zone = 100
+	/// This is some messy code, but emotes are like that. Anything in this list will not be triggered by the force-emote function.
+	var/list/emote_blacklist = list(
+		"custom", "customv", "customh", "me", "give", "help", "listbasic", "listtarget", "list", "suicide", "uguu", "juggle", "airquote", "airquotes",
+		"faint", "deathgasp", "collapse", "trip", "monologue", "miranda"
+	)
+
+	New()
+		. = ..()
+		var/datum/reagent/R = pick(concrete_typesof(/datum/reagent/fooddrink/alcoholic))
+		src.passkey = lowertext(replacetext(replacetext(R.name, " ", "_"), "'", ""))
+		if (!src.passkey)
+			src.passkey = "IMP-[rand(111, 999)]"
+		// The `uses_radio` variable only adds a sender component, not a two-way one. So we have to do that manually!
+		if (!src.net_id)
+			src.net_id = generate_net_id(src)
+		MAKE_DEFAULT_RADIO_PACKET_COMPONENT(src.net_id, null, src.pda_alert_frequency)
+		processing_items.Add(src)
+
+	disposing()
+		processing_items.Remove(src)
+		..()
+
+	deactivate()
+		..()
+		processing_items.Remove(src)
+
+	process()
+		src.adjust_heat(-1)
+		src.heat_dissipation = min(3, src.heat_dissipation + 0.05)
+
+	can_implant(mob/target, mob/user)
+		if (istype(target, /mob/living/critter/wraith/trickster_puppet))
+			boutput(user, SPAN_ALERT("The implanter shows an error on its little LCD screen: \"TARGET IS SECRETLY A GHOST\". Huh. \
+			You didn't know it even checked for that!"))
+			return
+		return ..()
+
+	implanted(mob/M, mob/I)
+		. = ..()
+		if (!src.burned_out)
+			// this is an anti-frustration feature with the goal of both making it easier to reference chat logs to know who is implanted with what,
+			// as well as to make sure players can still use packets if they forget to scan it onto their remote and didn't write down the network data
+			boutput(I, SPAN_NOTICE("You make a mental note that this implant's network ID is <b>[src.net_id]</b> and its passkey is <b>[src.passkey]</b>."))
+
+	receive_signal(datum/signal/signal)
+		if (src.burned_out || ON_COOLDOWN(src, "activate", 1 SECOND))
+			return
+		if (!signal || signal.encryption)
+			return
+		if (signal.data["address_1"] != src.net_id)
+			return
+		if (signal.data["sender"] == src.net_id)
+			return
+
+		var/command = signal.data["command"]
+		if (command == "ping")
+			var/datum/signal/ping_reply = get_free_signal()
+			ping_reply.source = src
+			ping_reply.data["device"] = "IMP_MARIONETTE"
+			ping_reply.data["sender"] = src.net_id
+			ping_reply.data["address_1"] = signal.data["sender"]
+			ping_reply.data["command"] = "ping_reply"
+			ping_reply.data["passkey"] = src.passkey
```

Comment:
Isn't the passkey meant to be a secret? If it responds with the passkey on ping then any random packetnerd could control it!

Reply:
that's a good point... I like the idea of having the ability to get the passkey from the ping, but I don't want it to be too easy to control. do you think that making a close proximity requirement (ala 3 tiles) to provide the passkey would fix that?

Reply:
Well if you can get the passkey from examining it and you get it put in chat when you implant it then that's probably good enough on its own? You could also add it to the notes verb ie "You implanted a marionette implant with passkey 'xyz'"
What's the benefit to having it available from being pinged besides other people being able to take control of it?

Reply:
honestly, I mostly just thought it'd be a neat facet to have. I also wanted it to be consistent with other packet-controlled systems, where you can run pings or other commands to receive the relevant access pass/key ID/etc. still, I suppose this is a bit of a different case since it's an antag item... I'll probably scrap it in light of that since, yeah, fair enough